extensions/pkg/terraformer/config.go

```diff
@@ -111,11 +111,16 @@ func (t *terraformer) initializerConfig(ctx context.Context) *InitializerConfig
 // Initializer to correctly create all the resources as specified in the given InitializerConfig.
 // A default implementation can be found in DefaultInitializer.
 func (t *terraformer) InitializeWith(ctx context.Context, initializer Initializer) Terraformer {
-	if err := initializer.Initialize(ctx, t.initializerConfig(ctx), t.ownerRef); err != nil {
+	config := t.initializerConfig(ctx)
+
+	if err := initializer.Initialize(ctx, config, t.ownerRef); err != nil {
 		t.logger.Error(err, "Could not create Terraformer ConfigMaps/Secrets")
 		return t
 	}
-	t.configurationDefined = true
+
+	t.configurationInitialized = true
+	t.stateInitialized = config.InitializeState
```

Comment:
I find confusing having `configurationInitialized` and `stateInitialized` in the `InitializeWith` func. Currently on infrastructure deletion in the provider extensions, in most of the cases we don't run `InitializeWith`. For example in provider-aws we only call `newTerraformer` - https://github.com/gardener/gardener-extension-provider-aws/blob/4b402bd01068b6bbd0a31b188bc2d16cd2b91d19/pkg/controller/infrastructure/actuator_delete.go#L49 , there is no call to `InitializeWith` func before we call `tf.Destroy`.
With this PR it currently works somehow, but I think that these `configurationInitialized` and `stateInitialized` fields are always set to `false` when `InitializeWith` is not called on deletion (most of the cases for provider extensions). Then these fields will be `false` when the actual state in the system is not `false`.

Reply:
Do you have a suggestion? E.g., where to move it?

Reply:
I think it should be in the `execute` func. Do we need actually this check for `!(t.configurationInitialized && t.stateInitialized)`? Can't we always check for the existing resources (as we do currently in the master) and simply tolerate if not all resources exists on destroy?

Reply:
> but I think that these configurationInitialized and stateInitialized fields are always set to false when InitializeWith is not called on deletion (most of the cases for provider extensions).
> [...]
> I think it should be in the execute func.

I think that setting these attributes in the `InitializeWith` function is correct. From the doc string:

```golang
// InitializeWith initializes the Terraformer with the given Initializer. It is expected from the
// Initializer to correctly create all the resources as specified in the given InitializerConfig.
// [...]
```

So, only when the initializer is called then all configuration resources are created/"initialized". It's not related to the `execute` function.

> Then these fields will be false when the actual state in the system is not false.

Yes, but this is correct, isn't it? These fields do not indicate whether the configuration resources in the system are initialized or not. They only indicate that the current instance of `Terraformer` has initialized them.

> Do we need actually this check for !(t.configurationInitialized && t.stateInitialized)? Can't we always check for the existing resources (as we do currently in the master) and simply tolerate if not all resources exists on destroy? Or better - can we simply skip the check for existing resources on destroy?

Yes, while working on this PR I was also thinking into this direction (completely removing this check - not only for the 'destroy' case). However, I decided to keep it because it's actually helpful and prevents starting a TF pod if the prerequisites are not fulfilled. It would also be okay without the check (we would maybe start a TF pod and it would fail, so the error message wouldn't be too helpful).
Anyways, I don't have a very strong opinion here. I slightly prefer to keep it as it is now. Alternatively, the entire check should be removed (not only for the 'destroy' case).

/cc @timebertt

Reply:
I have no strong opinion here.